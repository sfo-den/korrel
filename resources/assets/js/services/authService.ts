import { merge } from 'lodash'
import { http, localStorageService } from '@/services'
import { userStore } from '@/stores'

export interface UpdateCurrentProfileData {
  current_password: string | null
  name: string
  email: string
  avatar?: string
  new_password?: string
}

interface CompositeToken {
  'audio-token': string
  'token': string
}

const API_TOKEN_STORAGE_KEY = 'api-token'
const AUDIO_TOKEN_STORAGE_KEY = 'audio-token'

export const authService = {
  async login (email: string, password: string) {
    const token = await http.post<CompositeToken>('me', { email, password })

    this.setAudioToken(token['audio-token'])
    this.setApiToken(token.token)
  },

  async logout () {
    await http.delete('me')
    this.destroy()
  },

  getProfile: async () => await http.get<User>('me'),

  updateProfile: async (data: UpdateCurrentProfileData) => {
    merge(userStore.current, (await http.put<User>('me', data)))
  },

  getApiToken: () => localStorageService.get(API_TOKEN_STORAGE_KEY),

  hasApiToken () {
    return Boolean(this.getApiToken())
  },

  setApiToken: (token: string) => localStorageService.set(API_TOKEN_STORAGE_KEY, token),

  destroy: () => {
    localStorageService.remove(API_TOKEN_STORAGE_KEY)
    localStorageService.remove(AUDIO_TOKEN_STORAGE_KEY)
  },

  setAudioToken: (token: string) => localStorageService.set(AUDIO_TOKEN_STORAGE_KEY, token),

  getAudioToken: () => {
    // for backward compatibility, we first try to get the audio token, and fall back to the (full-privileged) API token
    return localStorageService.get(AUDIO_TOKEN_STORAGE_KEY) || localStorageService.get(API_TOKEN_STORAGE_KEY)
  }
}
