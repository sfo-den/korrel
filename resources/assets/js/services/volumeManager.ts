import { ref } from 'vue'
import { preferenceStore } from '@/stores'

export class VolumeManager {
  private input!: HTMLInputElement
  public volume = ref(0)

  public init (input: HTMLInputElement) {
    this.input = input
    this.set(preferenceStore.volume)
  }

  public get () {
    return this.volume.value
  }

  public set (volume: number) {
    this.volume.value = volume
    this.input.value = String(volume)
  }

  public mute () {
    this.set(0)
  }

  public unmute () {
    this.set(preferenceStore.volume)
  }
}

export const volumeManager = new VolumeManager()
