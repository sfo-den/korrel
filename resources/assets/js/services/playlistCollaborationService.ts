import { http } from '@/services'

export const playlistCollaborationService = {
  async createInviteLink (playlist: Playlist) {
    if (playlist.is_smart) {
      throw Error('Smart playlists are not collaborative.')
    }

    const token = (await http.post<{ token: string }>(`playlists/${playlist.id}/collaborators/invite`)).token
    return `${window.location.origin}/#/playlist/collaborate/${token}`
  },

  async acceptInvite (token: string) {
    return http.post<Playlist>(`playlists/collaborators/accept`, { token })
  }
}
