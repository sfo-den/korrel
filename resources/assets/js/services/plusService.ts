import { http } from '@/services'

export const plusService = {
  activateLicense: async (key: string) => {
    return await http.post('licenses/activate', { key })
  }
}
