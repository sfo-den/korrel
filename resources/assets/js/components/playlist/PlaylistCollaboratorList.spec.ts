import { expect, it } from 'vitest'
import UnitTestCase from '@/__tests__/UnitTestCase'
import factory from '@/__tests__/factory'
import { playlistCollaborationService } from '@/services'
import Component from './PlaylistCollaboratorList.vue'

new class extends UnitTestCase {
  private async renderComponent (playlist: Playlist) {
    const rendered = this.render(Component, {
      props: {
        playlist
      },
      global: {
        stubs: {
          ListItem: this.stub('ListItem')
        }
      }
    })

    await this.tick(2)

    return rendered
  }

  protected test () {
    it('renders', async () => {
      const playlist = factory<Playlist>('playlist', {
        is_collaborative: true
      })

      const fetchMock = this.mock(playlistCollaborationService, 'fetchCollaborators').mockResolvedValue(
        factory<PlaylistCollaborator>('playlist-collaborator', 5)
      )

      const { html } = await this.be().renderComponent(playlist)
      expect(fetchMock).toHaveBeenCalledWith(playlist)
      expect(html()).toMatchSnapshot()
    })
  }
}
