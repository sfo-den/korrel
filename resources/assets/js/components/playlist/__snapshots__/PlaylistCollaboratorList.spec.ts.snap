// Vitest Snapshot v1

exports[`renders 1`] = `<ul data-v-fa3dce2e=""><br data-v-fa3dce2e="" data-testid="ListItem" is="li" role="contributor" manageable="false" removable="false" collaborator="[object Object]"><br data-v-fa3dce2e="" data-testid="ListItem" is="li" role="contributor" manageable="false" removable="false" collaborator="[object Object]"><br data-v-fa3dce2e="" data-testid="ListItem" is="li" role="contributor" manageable="false" removable="false" collaborator="[object Object]"><br data-v-fa3dce2e="" data-testid="ListItem" is="li" role="contributor" manageable="false" removable="false" collaborator="[object Object]"><br data-v-fa3dce2e="" data-testid="ListItem" is="li" role="contributor" manageable="false" removable="false" collaborator="[object Object]"></ul>`;
